link/types.proto

```diff
@@ -0,0 +1,30 @@
+syntax = "proto3";
+
+package neo.fs.v2.link;
+
+option go_package = "github.com/nspcc-dev/neofs-api-go/v2/link/grpc;link";
+option csharp_namespace = "Neo.FileStorage.API.Link";
+
+import "refs/types.proto";
+
+// Link is a payload of helper objects that contain the full list of the split
+// chain objects' IDs. It is created only after the whole split chain is known
+// and signed. This object is the only object that refers to every "child object"
+// ID. It is NOT required for the original object assembling. It MUST have ALL
+// the "child objects" IDs. Child objects MUST be ordered according to the
+// original payload split, meaning the first payload part holder MUST be placed
+// at the first place in the corresponding link object. Sizes MUST NOT be omitted
```

Comment:
if we wanna use it, we should be able to rely on it so there is MUST word. but then, we need to validate it and that is N head requests per link object PUT (N is num of objects in the chain)

Reply:
> N head requests per link object

the problem is more specific - when assembling a split object, the list of descriptors of child objects can diverge with the stored children (not only the sizes btw, IDs too). The proposed validation is correct, but since we cannot guarantee it at the protocol level (this is a server implementation detail), the assembler - one who operates with split chain - must be prepared for the divergence in any case

protocol-level solution is to additionally store such metadata in the child object headers themselves (mentioned [here](https://github.com/nspcc-dev/neofs-api/issues/264#issuecomment-1908432474)). Then the link object will remain a pure stand-alone helper

Reply:
> but since we cannot guarantee it at the protocol level (this is a server implementation detail)

what do you mean? nodes may not accept such link objects

> must be prepared for the divergence in any case

there is no solution for #264 then. if you cannot trust the link object, you cannot be sure you are answering with a correct range

> protocol-level solution is to additionally store such metadata in the child object headers themselves

should also be validated somehow. a node can get Nth object with some offset in it. what does it do? just believes it is true? searches for the previous ones and calculates their sizes?

@roman-khimov

Reply:
> what do you mean? nodes may not accept such link objects

nodes may do whatever they want. They may accept linking objects w/o any validation

> there is no solution for https://github.com/nspcc-dev/neofs-api/issues/264 then

i dont deny current solution, im just highligting that linking object _can_ be unavailable or diverge with the split chain and any network node must be ready for this. In these cases, extra metadata in child objects _may_ be a painkiller in some cases

> what does it do?

its own implementation detail, each node can only control what to do itself and what it receives from other nodes

---

or maybe i misunderstood ur original comment. I dont see any problem with current structure and requirements. If u didnt mean that `N head requests per link object PUT` is a _problem_, then everything is ok to me

Reply:
> nodes may do whatever they want
>> what does it do?
>
> its own implementation detail, each node can only control what to do itself and what it receives from other nodes

i meant we need to find a balance to implement our "perfect" node that may exist in the world of "worst" nodes. if we may not rely on link object, i think the whole work is useless. but if we want to rely on it, we should validate it somehow

> extra metadata in child objects may be a painkiller in some cases

here i want to say that ensuring child object data is ok is the same to ensuring similar info about the link object

> If u didnt mean that N head requests per link object PUT is a problem, then everything is ok to me

yes. this and the consequences

Reply:
Validation should be implemented. Then I'd be optimistic wrt the link object contents. If it's not correct --- either you get an error (trying to access inexisting object) or reply with a wrong result, but it's not node's fault, garbage in -- garbage out.

Reply:
> Validation should be implemented.

Yes, sure. Just not sure about some link object replication: imagine you have a 1к parts object and receive a link to it for the first time: you need to do 1k head objects before you finish with this object. And you may be an attacker that sends 1k of such 1k-parts objects. 64mb data for a HEAD chaos in the NeoFS network.